service/history/execution/mutable_state_util_test.go

```diff
@@ -316,3 +327,252 @@ func TestConvertWorkflowRequests(t *testing.T) {
+		assert.NotNil(t, outputs)
+		assert.Equal(t, 1, len(outputs))
+		assert.Equal(t, key, outputs[0])
+	})
+
+	t.Run("convertUpdateChildExecutionInfos", func(t *testing.T) {
+		key := int64(0)
+		inputs := map[int64]*persistence.ChildExecutionInfo{key: {}}
+		outputs := convertUpdateChildExecutionInfos(inputs)
+		assert.NotNil(t, outputs)
+		assert.Equal(t, 1, len(outputs))
+		assert.Equal(t, inputs[key], outputs[0])
+	})
+
+	t.Run("convertUpdateSignalInfos", func(t *testing.T) {
+		key := int64(0)
+		inputs := map[int64]*persistence.SignalInfo{key: {}}
+		outputs := convertUpdateSignalInfos(inputs)
+		assert.NotNil(t, outputs)
+		assert.Equal(t, 1, len(outputs))
+		assert.Equal(t, inputs[key], outputs[0])
+	})
+}
+
+func TestScheduleDecision(t *testing.T) {
+	t.Run("mutable state has pending decision", func(t *testing.T) {
+		mockMutableState := NewMockMutableState(gomock.NewController(t))
+		mockMutableState.EXPECT().HasPendingDecision().Return(true)
+		err := ScheduleDecision(mockMutableState)
+		assert.Nil(t, err)
+	})
+
+	t.Run("internal service error", func(t *testing.T) {
+		mockMutableState := NewMockMutableState(gomock.NewController(t))
+		mockMutableState.EXPECT().HasPendingDecision().Return(false)
+		mockMutableState.EXPECT().AddDecisionTaskScheduledEvent(false).Return(nil, errors.New("some error"))
+		err := ScheduleDecision(mockMutableState)
+		assert.NotNil(t, err)
+		assert.Equal(t, "Failed to add decision scheduled event.", err.Error())
+	})
+
+	t.Run("success", func(t *testing.T) {
+		mockMutableState := NewMockMutableState(gomock.NewController(t))
+		mockMutableState.EXPECT().HasPendingDecision().Return(false)
+		mockMutableState.EXPECT().AddDecisionTaskScheduledEvent(false).Return(nil, nil)
+		err := ScheduleDecision(mockMutableState)
+		assert.Nil(t, err)
+	})
+}
+
+func TestFailDecision(t *testing.T) {
+	t.Run("failure", func(t *testing.T) {
+		mockMutableState := NewMockMutableState(gomock.NewController(t))
+		decision := &DecisionInfo{}
+		failureCause := new(types.DecisionTaskFailedCause)
+		mockMutableState.EXPECT().AddDecisionTaskFailedEvent(
+			decision.ScheduleID,
+			decision.StartedID,
+			*failureCause,
+			nil,
+			IdentityHistoryService,
+			"",
+			"",
+			"",
+			"",
+			int64(0),
+			"",
+		).Return(nil, errors.New("some error adding failed event"))
+		err := FailDecision(mockMutableState, decision, *failureCause)
+		assert.NotNil(t, err)
+		assert.Equal(t, "some error adding failed event", err.Error())
+	})
+
+	t.Run("success", func(t *testing.T) {
+		mockMutableState := NewMockMutableState(gomock.NewController(t))
+		decision := &DecisionInfo{}
+		failureCause := new(types.DecisionTaskFailedCause)
+		mockMutableState.EXPECT().AddDecisionTaskFailedEvent(
+			decision.ScheduleID,
+			decision.StartedID,
+			*failureCause,
+			nil,
+			IdentityHistoryService,
+			"",
+			"",
+			"",
+			"",
+			int64(0),
+			"",
+		).Return(nil, nil)
+		mockMutableState.EXPECT().FlushBufferedEvents()
```

Comment:
nit: it is better to havea comment that `FlushBufferedEvents` will only be called on success. Otherwise it is not clear what is the difference with previous.